specification.md

```diff
@@ -1574,3 +1579,119 @@ to send over the connection. When using a single connection to transmit
 larger messages, TCP buffers can be more efficiently utilized.
 Additionally, it becomes feasible to tune buffer sizes in order to
 optimize the throughput of connections to particular services.
+
+
+
+### Appendix C - examples
+
+This Appendix contains some examples of how the API can be used. Unlike the
+specification language, the examples are in a hypothetical Python binding, which
+is expected to be a relatively frequently used binding.
+
+#### Submit and wait for N jobs
+
+This example shows how to submit `N` jobs and synchronously wait for them to
+complete.
+
+```python
+import psij
+
+jex = psij.JobExectorFactory.get_instance('slurm')
+
+def make_job():
+    job = psij.Job()
+    spec = psij.JobSpecification()
+    spec.executable = '/bin/sleep'
+    spec.arguments = ['10']
+    job.specification = spec
+    return job
+
+jobs = []
+for i in range(N):
+    job = make_job()
+    jobs.append(job)
+    jex.submit(job)
+
+for i in range(N):
```

Comment:
Hmm, I really think that we should make the full API bulk-compatible:
```py
jobs = [make_job() for _ in range(N)
jex.submit(jobs)
jex.wait(jobs)
```

Reply:
Perhaps I should more clearly state what I think about the bulk API: nothing that we discussed so far suggests a reason to have it except that it may make an efficient implementation easier to implement, while making the interface more complex. Everything else points to not having a bulk API. I do not think having a bulk API is, therefore, justified.

Reply:
Thanks for the clarity of the statement, that helps :-D  Let me follow suit:  I think bulks are essential on the API level.  Apart from that, they are also easy to use and make code simpler.

Seems we have some work to do :-P

Why do you think they make the interface more complex?  I am puzzled by that statement:
```py
jobs = [make_job() for _ in range(N)]
for job in jobs:
    jex.submit(job)

for job in jobs:
    job.wait()
```
vs
```py
jobs = [make_job() for _ in range(N)]
jex.submit(jobs)
jex.wait(jobs)
```

The loops are obvious, but the first also lacks symmetry (submit is on `jex`, wait is on `job`).  I mean, both are kinda simple of course - but I don't see how bulks add complexity.

Also: the point of bulks is not to make the implementation simpler (it does not), but to make an efficient implementation possible.

Reply:
> Thanks for the clarity of the statement, that helps :-D Let me follow suit: I think bulks are essential on the API level. Apart from that, they are also easy to use and make code simpler.
> 
> Seems we have some work to do :-P

I say let's settle this once and for all... outside... in Issue #55 :)

> 
> Why do you think they make the interface more complex? 

Because every time the user needs to submit jobs, they need to decide whether to use the bulk or the non-bulk call. If the reason for having a bulk call is that implementations will possibly have better performance if the bulks are used, users will strive to use the bulk call even if they need to submit one job at a time. As I have stated before, we are bleeding implementation details into the API, and we are increasing the cognitive load on the users, but we do so without providing enough information so that the users can make an informed decision. Is it worth it to a user to queue single jobs only to use the bulk call? That can only be answered if the implementation is clear on how efficient the bulk call is vs. the non-bulk call. And it cannot really be answered in general, for all implementations. So, ultimately, our user will probably decide to err on the side of caution and always call the bulk API, implementing a queue if necessary to bulk up single jobs. But this implies that we are failing at design. If we know that this is very likely to happen, we should offset the effort from the user and do this bulking ourselves, so that the bulking code only gets written once. I mean, this is fundamentally what we are doing in this project: writing this API and adapters to that users don't have to implement their own every time they have to submit jobs in a portable way.

I'll leave bulk exception handling out, but that's an additional complication. It is pretty clear in the current API. The existence of the FaultDetail class is solely due to the bulk exceptions.

>I am puzzled by that statement:
> 
> ```python
> jobs = [make_job() for _ in range(N)]
> for job in jobs:
>     jex.submit(job)
> 
> for job in jobs:
>     job.wait()
> ```
> 
> vs
> 
> ```python
> jobs = [make_job() for _ in range(N)]
> jex.submit(jobs)
> jex.wait(jobs)
> ```
> 
> The loops are obvious, but the first also lacks symmetry (submit is on `jex`, wait is on `job`). I mean, both are kinda simple of course - but I don't see how bulks add complexity.

I would like to point out that one can have this argument for almost everything: add bulk versions so that users don't have to write for loops.

> 
> Also: the point of bulks is not to make the implementation simpler (it does not), but to make an efficient implementation possible.

I think the bulk submission appendix goes over this. One can queue jobs and periodically drain the queue into an internal bulk submit without exposing that to the user. But there are other ways to do things that can lead to even better performance. An efficient implementation is possible with or without the public bulk methods. 

The way bulks can make an implementation easier is by avoiding the aforementioned queue. But it's doing so at the expense of the user who might now have to do write that code themselves.

Reply:
Resolving since we moved discussion to #55